src/olympia/devhub/tests/test_forms.py

```diff
@@ -943,95 +943,169 @@ def test_site_permission_generator_origin_valid(origin):
-    request = req_factory_factory('/', user=user_factory())
-    addon = addon_factory(type=amo.ADDON_SITE_PERMISSION, users=[request.user])
-    version = addon.versions.get()
-    FileSitePermission.objects.create(
-        file=version.file, permissions=_DEFAULT_SITE_PERMISSIONS
-    )
-    version.installorigin_set.create(origin='https://example.com')
+    def test_site_permission_generator_duplicates_same_permission(self):
+        addon = addon_factory(type=amo.ADDON_SITE_PERMISSION, users=[self.request.user])
+        version = addon.versions.get()
+        FileSitePermission.objects.create(
+            file=version.file, permissions=_DEFAULT_SITE_PERMISSIONS
+        )
+        version.installorigin_set.create(origin='https://example.com')
 
-    # User already has a site permission, for the same set of permissions, but
-    # for a different origin.
-    form = forms.SitePermissionGeneratorForm(
-        {'site_permissions': _DEFAULT_SITE_PERMISSIONS, 'origin': 'https://foo.com'},
-        request=request,
-    )
-    assert form.is_valid()
+        # User already has a site permission, for the same set of permissions, but
+        # for a different origin.
+        form = forms.SitePermissionGeneratorForm(
+            {
+                'site_permissions': _DEFAULT_SITE_PERMISSIONS,
+                'origin': 'https://foo.com',
+            },
+            request=self.request,
+        )
+        assert form.is_valid()
 
-    # Duplicate.
-    version.installorigin_set.create(origin='https://foo.com')
-    form = forms.SitePermissionGeneratorForm(
-        {'site_permissions': _DEFAULT_SITE_PERMISSIONS, 'origin': 'https://foo.com'},
-        request=request,
-    )
-    assert not form.is_valid()
+        # Duplicate.
+        version.installorigin_set.create(origin='https://foo.com')
+        form = forms.SitePermissionGeneratorForm(
+            {
+                'site_permissions': _DEFAULT_SITE_PERMISSIONS,
+                'origin': 'https://foo.com',
+            },
+            request=self.request,
+        )
+        assert not form.is_valid()
 
-    # Deleting the duplicate should make submission possible again.
-    addon.delete()
-    form = forms.SitePermissionGeneratorForm(
-        {'site_permissions': _DEFAULT_SITE_PERMISSIONS, 'origin': 'https://foo.com'},
-        request=request,
-    )
-    assert form.is_valid()
+        # Deleting the duplicate should make submission possible again.
+        addon.delete()
+        form = forms.SitePermissionGeneratorForm(
+            {
+                'site_permissions': _DEFAULT_SITE_PERMISSIONS,
+                'origin': 'https://foo.com',
+            },
+            request=self.request,
+        )
+        assert form.is_valid()
+
+    def test_rate_limiting(self):
+        self.request.META['REMOTE_ADDR'] = '5.6.7.8'
+        with freeze_time('2021-01-08 15:16:23.42') as frozen_time:
+            for x in range(0, 6):
```

Comment:
Is this just to make this happen 7 times?

Reply:
Yeah (copied from another test above). It would be nice to have a test method to set that up in a cleaner way someday...